File: Paintroid/src/org/catrobat/paintroid/command/implementation/FillCommand.java
/**
 *  Catroid: An on-device visual programming system for Android devices
 *  Copyright (C) 2010-2012 The Catrobat Team
 *  (<http://developer.catrobat.org/credits>)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  http://www.catroid.org/catroid/licenseadditionalterm
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.catrobat.paintroid.command.implementation;

import org.catrobat.paintroid.tools.helper.floodfill.QueueLinearFloodFiller;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Point;

public class FillCommand extends BaseCommand {

	private static final float SELECTION_THRESHOLD = 50.0f;
	private Point mClickedPixel;

	public FillCommand(Point clickedPixel, Paint currentPaint) {
		super(currentPaint);
		mClickedPixel = clickedPixel;
	}

	@Override
	public void run(Canvas canvas, Bitmap bitmap) {
		setChanged();
		notifyStatus(NOTIFY_STATES.COMMAND_STARTED);
		if (mClickedPixel == null) {
			setChanged();
			notifyStatus(NOTIFY_STATES.COMMAND_FAILED);
			return;
		}

		int colorToReplace = bitmap.getPixel(mClickedPixel.x, mClickedPixel.y);

		int pixels[] = new int[bitmap.getWidth() * bitmap.getHeight()];
		bitmap.getPixels(pixels, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(),
				bitmap.getHeight());

		QueueLinearFloodFiller.floodFill(pixels, bitmap.getWidth(),
				bitmap.getHeight(), mClickedPixel, colorToReplace,
				mPaint.getColor(), SELECTION_THRESHOLD);

		bitmap.setPixels(pixels, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(),
				bitmap.getHeight());

		setChanged();
		notifyStatus(NOTIFY_STATES.COMMAND_DONE);
	}
}

File: Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/FillCommand.kt
/*
 * Paintroid: An image manipulation application for Android.
 *  Copyright (C) 2010-2022 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.command.implementation

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Point
import org.catrobat.paintroid.command.Command
import org.catrobat.paintroid.contract.LayerContracts
import org.catrobat.paintroid.tools.helper.FillAlgorithmFactory

class FillCommand(private val fillAlgorithmFactory: FillAlgorithmFactory, clickedPixel: Point, paint: Paint, colorTolerance: Float) : Command {

    var clickedPixel = clickedPixel; private set
    var paint = paint; private set
    var colorTolerance = colorTolerance; private set

    override fun run(canvas: Canvas, layerModel: LayerContracts.Model) {
        val currentLayer = layerModel.currentLayer
        currentLayer ?: return
        currentLayer.bitmap.let { bitmap ->
            val colorToBeReplaced = bitmap.getPixel(clickedPixel.x, clickedPixel.y)
            val fillAlgorithm = fillAlgorithmFactory.createFillAlgorithm()
            fillAlgorithm.setParameters(
                    bitmap,
                    clickedPixel,
                    paint.color,
                    colorToBeReplaced,
                    colorTolerance
            )
            fillAlgorithm.performFilling()
        }
    }

    override fun freeResources() {
        // No resources to free
    }
}
